File: server/src/llm/model_info/ollama_utils.py
import requests
from aif_types.ollama import OllamaModelInfo
from llm.llm_uri_utils import remove_ollama_tag_latest
from server_config import serverConfig


def get_host() -> str:
    """
    Ollama is running in the host machine, however, we prefer host.docker.internal because we only recommend running the server within a Docker container.
    """
    return "http://localhost:11434" if serverConfig.useLocalServer else "http://host.docker.internal:11434"


def health_check() -> bool:
    url = f"{get_host()}"
    try:
        response = requests.get(url)
    except Exception:
        return False
    return response.status_code == 200


def get_local_model_map() -> dict[str, OllamaModelInfo]:
    url = f"{get_host()}/api/tags"
    try:
        response = requests.get(url)
        if response.status_code == 200:
            json_response = response.json()
            ollama_model_info_map = {}
            for model_info_json in json_response["models"]:
                id = remove_ollama_tag_latest(model_info_json["model"])
                model_info = OllamaModelInfo(
                    name=model_info_json["name"],
                    id=id,
                    size=model_info_json["size"],
                    modified_at=model_info_json["modified_at"],
                    version=None
                )

                ollama_model_info_map[id] = model_info
            return ollama_model_info_map
        else:
            raise Exception(f"Failed to get Ollama data from {url}")
    except Exception as e:
        raise e


def download_model(model_name: str) -> bool:
    url = f"{get_host()}/api/pull"
    try:
        response = requests.post(url, json={ "model": model_name})
        return response.status_code == 200
    except Exception:
        return False

File: packages/server-js/src/utils/OllamaUtils.ts
import ServerConfig from "../config/ServerConfig";
import { ApiOutStream } from "../types/ApiOutStream";

namespace OllamaUtils {
    export function getHost(): string {
        // Case 1: when the server is running in Docker, as Ollama is running in the host machine, host is: `host.docker.internal`
        // Case 2: for some reason, `127.0.0.1` can work but `localhost` cannot
        // return ServerConfig.useLocalServer ? "http://localhost:11434" : "http://host.docker.internal:11434";
        return ServerConfig.useLocalServer ? "http://127.0.0.1:11434" : "http://host.docker.internal:11434";
    }

    export async function isHealthy(): Promise<boolean> {
        try {
            const result = await fetch(getHost());
            return result.status === 200;
        } catch (error) {
            return false;
        }
    }

    export function startOllamaServer(out: ApiOutStream): void {
        isHealthy().then((healthy: boolean) => {
            if (healthy) {
                out.end();
            } else {
                out.write("Ollama server is not running. Starting Ollama server...");
                _startOllamaServer(out);
            }
        });
    }

    export async function listDownloadedModels(): Promise<string[]> {
        const endpoint = `${getHost()}/api/tags`;
        const response = await (await fetch(endpoint)).json();
        return (response as any).models.map((model) => model.name);
    }

    export function downloadModel(modelName: string, out: ApiOutStream): void {
        const endpoint = `${getHost()}/api/pull`;
        fetch(endpoint, {
            method: "POST",
            body: JSON.stringify({ name: modelName, stream: true }),
        }).then((response) => {
            if (!response || response.status !== 200) {
                out.error(`Failed to download model ${modelName}`);
                return null;
            } else {
                return response.text();
            }
        }).then((text) => {
            if (text !== null) {
                out.end();
            }
        });
    }

    export function deleteModel(modelName: string, out: ApiOutStream): void {
        const endpoint = `${getHost()}/api/delete`;
        fetch(endpoint, {
            method: "DELETE",
            body: JSON.stringify({ name: modelName }),
        }).then((response) => {
            if (!response || response.status !== 200) {
                out.error(`Failed to delete model ${modelName}`);
                return null;
            } else {
                return response.text();
            }
        }).then((text) => {
            if (text !== null) {
                out.end();
            }
        });
    }
}

function _startOllamaServer(out: ApiOutStream): void {
    let command: string | undefined;
    
    if (process.platform === "win32") {
        out.write("\nError: not implemented");
    } else if (process.platform === "darwin") {
        command = `osascript -e 'tell app "Terminal"
do script "ollama serve"
end tell'`;
    } else if (process.platform === "linux") {
        out.write("\nError: not implemented");
    }

    const { exec } = require('child_process');
    exec(command, (error, stdout, stderr) => {
        if (error) {
            out.error(`Failed to start Ollama server: ${error}`);
        } else {
            out.write("Ollama server is running.", "success");
        }      
        out.end();
    });
}

export default OllamaUtils;
